startup-scripts.md

```diff
@@ -0,0 +1,305 @@
+# volume's userDataSecretRef field
+
+cat << END > my-vm.yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: VirtualMachine
+metadata:
+  name: myvm
+spec:
+  terminationGracePeriodSeconds: 5
+  domain:
+    resources:
+      requests:
+        memory: 64M
+    devices:
+      disks:
+      - name: registrydisk
+        volumeName: registryvolume
+        disk:
+          dev: vda
+      - name: cloudinitdisk
+        volumeName: cloudinitvolume
+        disk:
+          dev: vdb
+  volumes:
+    - name: registryvolume
+      registryDisk:
+        image: kubevirt/cirros-registry-disk-demo:devel
+    - name: cloudinitvolume
+      cloudInitNoCloud:
+        userDataSecretRef: my-vm-secret
+END
+
+# Post the secret first, and then post the VM
+kubectl create -f my-secret.yaml
+kubectl create -f my-vm.yaml
+```
+
+### Injecting SSH keys with Cloud-init's Cloud-config
+
+In the examples so far, the cloud-init userdata script has been a bash script.
+Cloud-init has it's own configuration that can handle some common tasks such as
+user creation and ssh key injection.
+
+More cloud-config examples can be found here. [Cloud-init Examples](https://cloudinit.readthedocs.io/en/latest/topics/examples.html)
+
+Below is an example of using cloud-config to inject an ssh key for the default
+user (fedora in this case) of a [Fedora Atomic](https://getfedora.org/en/atomic/download/) disk image.
+
+```
+# Create the cloud-init cloud-config userdata.
+cat << END > startup-script
+#cloud-config
+password: atomic
+chpasswd: { expire: False }
+ssh_pwauth: False
+ssh_authorized_keys:
+    - ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQC6zdgFiLr1uAK7PdcchDd+LseA5fEOcxCCt7TLlr7Mx6h8jUg+G+8L9JBNZuDzTZSF0dR7qwzdBBQjorAnZTmY3BhsKcFr8Gt4KMGrS6r3DNmGruP8GORvegdWZuXgASKVpXeI7nCIjRJwAaK1x+eGHwAWO9Z8ohcboHbLyffOoSZDSIuk2kRIc47+ENRjg0T6x2VRsqX27g6j4DfPKQZGk0zvXkZaYtr1e2tZgqTBWqZUloMJK8miQq6MktCKAS4VtPk0k7teQX57OGwD6D7uo4b+Cl8aYAAwhn0hc0C2USfbuVHgq88ESo2/+NwV4SQcl3sxCW21yGIjAGt4Hy7J fedora@localhost.localdomain
+END
+
+# Create the VM spec
+cat << END > my-vm.yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: VirtualMachine
+metadata:
+  name: sshvm
+spec:
+  terminationGracePeriodSeconds: 0
+  domain:
+    resources:
+      requests:
+        memory: 1024M
+    devices:
+      disks:
+      - name: registrydisk
+        volumeName: registryvolume
+        disk:
+          dev: vda
+      - name: cloudinitdisk
+        volumeName: cloudinitvolume
+        disk:
+          dev: vdb
+  volumes:
+    - name: registryvolume
+      registryDisk:
+        image: kubevirt/fedora-atomic-registry-disk-demo:devel
+    - name: cloudinitvolume
+      cloudInitNoCloud:
+        userDataBase64: $(cat startup-script | base64 -w0)
+END
+
+# Post the Virtual Machine spec to KubeVirt.
+kubectl create -f my-vm.yaml
+
+# Connect to VM with passwordless ssh key
+ssh -i <insert private key here> fedora@<insert ip here>
+
+```
+
+### Inject SSH key using Custom Shell Script
```

Comment:
From my point of view, having a *small* shell example and the ssh-key cloud-config example above is good enough. I think people will understand, that if you have a shell script run as root, that you can do arbitrary things with it. This example is pretty huge and does not really provide much benefit.

Reply:
What I noticed while testing all of this is that not all disk images interpreted/applied the cloud-init config correctly, which meant that some people are going to find the cloud-config way of injecting ssh won't work for them.

Since user creation and ssh key injection is such a prevalent use case, I thought it would be helpful to provide an alternative approach people could start with. 

What would you think about replacing the small shell script example with this one that does user/ssh-key injection?

Reply:
> This example is pretty huge

ha, it's only 8 lines. Same as the GCE example on this page. https://cloud.google.com/compute/docs/startupscript

Reply:
I don't know why I though that it is huge. Works for me like it is.